### PR TITLE
fix run_seq2seq.py; porting trainer tests to it

### PR DESCRIPTION
**Pull request by stas00**
This PR:

- restores some of the essential dropped functionality from `finetune_trainer.py` - I'm almost sure this is far far from complete since so much was just dropped 
- ports wmt_en_ro test data to `jsonlines` - I move the tests dataset into the root of examples so that it can be accessed by a variety of sub-projects.
- ports DeepSpeed tests to use `run_seq2seq.py`
- ports the other trainer script to use `run_seq2seq.py`

@sgugger

### CODE REVIEW

File: examples/seq2seq/run_seq2seq.py [hunk 1]
```diff
@@ -351,8 +374,15 @@ def main():
...
-    if model.config.decoder_start_token_id is None and isinstance(tokenizer, MBartTokenizer):
-        model.config.decoder_start_token_id = tokenizer.lang_code_to_id[data_args.target_lang]
+    if model.config.decoder_start_token_id is None and isinstance(tokenizer, (MBartTokenizer, MBartTokenizerFast)):
+        assert (
+            data_args.target_lang is not None and data_args.source_lang is not None
+        ), "mBart requires --target_lang and --source_lang"
+        if isinstance(tokenizer, MBartTokenizer):
+            model.config.decoder_start_token_id = tokenizer.lang_code_to_id[data_args.target_lang]
+        else:
+            model.config.decoder_start_token_id = tokenizer.convert_tokens_to_ids(data_args.target_lang)
```
Comment (sgugger) [inline examples/seq2seq/run_seq2seq.py:384]: Thanks for putting that logic back. @patil-suraj @patrickvonplaten since you better than me, shouldn't this be done by `set_tgt_lang_special_tokens` inside mBART?

**Review by sgugger — APPROVED** [review]
Thanks for all your work on this! The test looks perfect to me, just have a few comments on the script itself.

File: examples/seq2seq/run_seq2seq.py [hunk 2]
```diff
@@ -561,13 +591,10 @@ def compute_metrics(eval_preds):
         train_result = trainer.train(resume_from_checkpoint=checkpoint)
         trainer.save_model()  # Saves the tokenizer too for easy upload
 
-        output_train_file = os.path.join(training_args.output_dir, "train_results.txt")
+        metrics = train_result.metrics
+        metrics["train_n_objs"] = data_args.max_train_samples
```
[the remark this answers is not in the document]
Comment (stas00) [reply examples/seq2seq/run_seq2seq.py:595]: yes, I need this information when comparing results. it's too easy to git mixed up in the args being changed, so the important params should be bundled with the metrics IMHO.

i.e. it can be essential to know how much data was fed to the script to know when things go wrong. at least in my experience so far I often refer to this number when analyzing results.

File: examples/seq2seq/run_seq2seq.py [hunk 3]
```diff
@@ -244,6 +251,22 @@ def __post_init__(self):
...
+    Args:
+    - split: one of train, val, test
+    - metrics: metrics dict
+    - output_dir: where to save the metrics
+    """
+
+    logger.info(f"***** {split} metrics *****")
+    for key in sorted(metrics.keys()):
+        logger.info(f"  {key} = {metrics[key]}")
+    save_json(metrics, os.path.join(output_dir, f"{split}_results.json"))
```
[the remark this answers is not in the document]
Comment (stas00) [reply examples/seq2seq/run_seq2seq.py:267]: As I said. I did not refactor it but restored the original code which was in this way in first place.

I will rewrite it explicitly as you request.

File: examples/seq2seq/run_seq2seq.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stas00) [reply examples/seq2seq/run_seq2seq.py:595]: but you have an important point there - what if the total number of samples was less than max? could we report this in metrics?

[the remark this answers is not in the document]
Comment (sgugger) [reply examples/seq2seq/run_seq2seq.py:595]: It would be `min(data_args.max_train_samples, len(train_dataset))` in that case.

File: examples/seq2seq/run_seq2seq.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stas00) [reply examples/seq2seq/run_seq2seq.py]: Done

File: examples/seq2seq/run_seq2seq.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stas00) [reply examples/seq2seq/run_seq2seq.py]: Awesome - thank you for the tip and I renamed it too so it's consistent with `samples_per_sec`

File: examples/seq2seq/run_seq2seq.py (hunk 1 above)
Comment (stas00) [reply examples/seq2seq/run_seq2seq.py:368]: Can we make a priority on merging this, so that I could finish porting the other tests? And this can be dealt with separately? 

I can open an issue so that it doesn't fall between the cracks.

Thank you!

Comment (patil-suraj) [reply examples/seq2seq/run_seq2seq.py:368]: `set_tgt_lang_special_tokens` is a method on `MBartTokenizer` so this needs to be done outside of the model.

Also, IMO `MBartTokenizerFast` should also have the `lang_code_to_id` attribute, not sure why it was treated differently.

**Comment by stas00** [comment]
OK, I decided to go ahead and port the other scripts instead of waiting for merging of the first set. Had to make some more fixes in the script while at it.

File: examples/seq2seq/run_seq2seq.py (hunk 1 above)
Comment (stas00) [reply examples/seq2seq/run_seq2seq.py:368]: OK, I decided to go ahead and port the other scripts instead of waiting for merging of the first set. 

Had to make some more fixes in the script while at it. 

So no rush.

File: examples/tests/trainer/test_trainer_ext.py [hunk 4]
```diff
@@ -60,44 +63,46 @@ def require_apex(test_case):
         return test_case
 
 
-class TestFinetuneTrainer(TestCasePlus):
-    def finetune_trainer_quick(self, distributed=None, extra_args_str=None):
+class TestTrainerExt(TestCasePlus):
```
Comment (sgugger) [inline examples/tests/trainer/test_trainer_ext.py:66]: I don't get what `Ext` means? Is it for extended? I don't really care for the class name, but the filename should have the full name so `test_trainer_extended.py`.
It's also actually testing the `Seq2SeqTrainer` in practice.

**Review by sgugger — APPROVED** [review]
LGTM and we can merge and deal with the MBart thing later on (it's true it's something in the model, not the tokenizer, will think of what the best API could be).

Just have one naming nit.

Thanks for porting those!

File: examples/tests/trainer/test_trainer_ext.py (hunk 4 above)
Comment (stas00) [reply examples/tests/trainer/test_trainer_ext.py:66]: Well, just wanted to differentiate it from test_trainer. I think it'll go through a refinement later to probably separate seq2seq distributed tests from integration tests, as they are in one pile at the moment. This is just because they reuse the same helper methods. It will evolve and improve as we expand more tests.

I don't think the filename has to match class name - this is because we have multiple cases where the filename contains multiple class names.

Bottom line, this is a WIP, but I had to change the original `Finetune`... Totally your call, @sgugger, on how do you prefer I name it.

Comment (sgugger) [reply examples/tests/trainer/test_trainer_ext.py:66]: I don't mind if the filename doesn't match the class, it's just that `test_trainer_ext.py` is not very informative and I have no idea what's tested inside just by looking at it ;-)
Since we are in tests/trainer, the trainer part can be dropped, so there is plenty of space to be more informative. With your plans to separate, it could be `test_integration` and `test_distributed` once split.

Comment (stas00) [reply examples/tests/trainer/test_trainer_ext.py:66]: Yes, that's the idea for when the dust settles - there are at least 3 types of tests in there at the moment.

That is it will not be called `test_trainer_ext.py` for long, how about I called it `test_to_be_sorted.py` for now so it's loud and clear it's a pile of things.  Or can just rename it back to what it is now under `seq2seq`.

The point of this PR is to transition to the new tool to enable Suraj to move along. And I will work in future PRs to make things intuitive and named well. I will not forget.

Comment (sgugger) [reply examples/tests/trainer/test_trainer_ext.py:66]: Sounds good to me!

File: examples/seq2seq/run_seq2seq.py (hunk 1 above)
Comment (patrickvonplaten) [reply examples/seq2seq/run_seq2seq.py:368]: Agree with @patil-suraj here -> IMO `MBartTokenizerFast` should have the `lang_code_to_id` so that we don't need an if-else here...but this is not necessarily the responsibility of this PR, so it's fine for me as it is

Comment (patrickvonplaten) [reply examples/seq2seq/run_seq2seq.py:368]: Maybe for models such as MBart that don't have an unique `decoder_start_token_id`, we should maybe in the future just leave `config.decoder_start_token_id=None` and then throw an error / warning when generating

Comment (patrickvonplaten) [reply examples/seq2seq/run_seq2seq.py:368]: Setting it manually here `model.config.decoder_start_token_id` is the correct thing to do here, but not super pretty IMO -> it would be better to incentivise the user to either insert at init or when calling `generate` But because of backward comp we can't really change it now for mBART anyways...
